### ISSUE TITLE
slow function/gradient evaluation for complex-valued parameters

### ISSUE THREAD

**Issue by fmkroci** [google/jax#15735]
### Description

Hello everybody,

thank you very much for all your work on jax! 

I noticed a performance issue while writing a likelihood function. The issue is possibly a version of #3068.
The function looks something like this:
```
def dummyLikelihood(T, data):
	sumOfAmps = data @ T
	return np.sum(np.log(np.abs(sumOfAmps)**2))
```
[I attached a self-contained example at the end]

Where data is of shape (nmbEvents, nDim) and dtype is complex128 and T is of length nDim also of dtype complex128.
Typical values are nmbEvents ~ 100000; nDim ~ 1000. From my understanding the calculation should be mostly limited by memory bandwidth.

I noticed a slow down (for the jitted version of the function) as compared to an implementation using autograd or original numpy (both using openBLAS).
The same is true for the gradient/value_and_grad of the function.

The slow-down disappears as soon as I switch to float.

I found an implementation that is much faster (even than the original numpy implementation) by using complex dtypes only internally:
```
@jit
def dummyLikelihood_jnp_fast(T_real, data_re, data_im):
    T = T_real[::2] + 1.j*T_real[1::2]
    data = data_re + 1.j*data_im
    sumOfAmps = data @ T
    return jnp.sum(jnp.log(jnp.real(jnp.abs(sumOfAmps)**2)))
```
However, the gradient/value_and_grad of this function is then even slower than the original complex-valued implementation.

I got the following measurements with my example code:
```
timing likelihood numpy:
dtype is: complex128
average evaluation time (s):     0.0439

timing likelihood autograd:
dtype is: complex128
average evaluation time (s):     0.0433

timing likelihood jax:
dtype is: complex128
average evaluation time (s):     0.0644

timing likelihood numpy:
dtype is: float64
average evaluation time (s):     0.0214

timing likelihood autograd:
dtype is: float64
average evaluation time (s):     0.0213

timing likelihood jax:
dtype is: float64
average evaluation time (s):     0.0213

timing value_and_grad autograd:
dtype is: complex128
average evaluation time (s):     0.0877

timing value_and_grad jax:
dtype is: complex128
average evaluation time (s):     0.5855

timing value_and_grad autograd:
dtype is: float64
average evaluation time (s):     0.0447

timing value_and_grad jax:
dtype is: float64
average evaluation time (s):     0.0406

```

I hope that I am not missing something here.
Thank you in advance for your help.

Example:
```
# switch between float64 and float32
useFloat64 = True

# realisitic number of events
nmbEvents = 100000
# realistic number of dimensions
nDim = 800

import os
# disable multithreading by default: we do not want to allow openBLAS to be faster just by multithreading
os.environ['OPENBLAS_NUM_THREADS'] = "1"

from jax import jit
import jaxlib
import jax
from jax.config import config
config.update('jax_platform_name', 'cpu')
config.update("jax_enable_x64", useFloat64)

from jax import numpy as jnp
from autograd import numpy as anp
import numpy as onp

dtype=onp.float64 if useFloat64 else onp.float32

from jax import value_and_grad as jax_value_and_grad
from autograd import value_and_grad as autograd_value_and_grad

import time


def time_function(testName, fun, T, data, nIter=10):
    print("timing {0}:".format(testName))
    print("dtype is: {0}".format(str(T.dtype)))

    block = type(T[0]) is jaxlib.xla_extension.ArrayImpl

    # initial call to trigger compile for jax
    val = fun(T, data)
    if type(val) is tuple:
        isVgrad=True
    else:
        isVgrad=False

    if block:
        vblock = val[0] if isVgrad else val
        vblock.block_until_ready()  

    start = time.time()
    for i in range(nIter):
        val = fun(T, data)
        if block:
            vblock = val[0] if isVgrad else val
            vblock.block_until_ready()
    end = time.time()
    print("average evaluation time (s): {:10.4f}".format((end - start)/nIter))
    print()


### dummy likelihoods ###

def dummyLikelihood_numpy(T, data):
	sumOfAmps = data @ T
	return onp.sum(onp.log(onp.abs(sumOfAmps)**2))


def dummyLikelihood_autograd(T, data):
	sumOfAmps = data @ T
	return anp.sum(anp.log(anp.abs(sumOfAmps)**2))


dummyLikelihood_autograd_vgr = autograd_value_and_grad(dummyLikelihood_autograd)


@jit
def dummyLikelihood_jax(T, data):
    sumOfAmps = data @ T
    return jnp.sum(jnp.log(jnp.abs(sumOfAmps)**2))


dummyLikelihood_jax_vgr = jit(jax_value_and_grad(dummyLikelihood_jax))


# create pseudo-data
onp_data_re = onp.random.rand(nmbEvents,nDim).astype(dtype)
onp_data_im = onp.random.rand(nmbEvents,nDim).astype(dtype)
onp_data = onp_data_re + 1.j * onp_data_im

onp_T_re = onp.random.rand(nDim).astype(dtype)
onp_T_im = onp.random.rand(nDim).astype(dtype) 
onp_T = onp_T_re + 1.j*onp_T_im

# copy to autograd
anp_data    = anp.array(onp_data,       copy=True)
anp_T       = anp.array(onp_T,          copy=True)

anp_data_re    = anp.array(onp_data_re, copy=True)
anp_T_re       = anp.array(onp_T_re,    copy=True)

# copy to jax
jnp_data    = jnp.array(onp_data,       copy=True)
jnp_T       = jnp.array(onp_T,          copy=True)

jnp_data_re    = jnp.array(onp_data_re, copy=True)
jnp_T_re       = jnp.array(onp_T_re,    copy=True)


functionList = [
                    {'testName': 'likelihood numpy', 'fun': dummyLikelihood_numpy, 'T': onp_T, 'data': onp_data},
                    {'testName': 'likelihood autograd','fun': dummyLikelihood_autograd, 'T': anp_T, 'data': anp_data},
                    {'testName': 'likelihood jax', 'fun': dummyLikelihood_jax, 'T': jnp_T, 'data': jnp_data},

                    {'testName': 'likelihood numpy', 'fun': dummyLikelihood_numpy, 'T': onp_T_re, 'data': onp_data_re},
                    {'testName': 'likelihood autograd', 'fun': dummyLikelihood_autograd, 'T': anp_T_re, 'data': anp_data_re},
                    {'testName': 'likelihood jax', 'fun': dummyLikelihood_jax, 'T': jnp_T_re, 'data': jnp_data_re},

                    {'testName': 'value_and_grad autograd', 'fun': dummyLikelihood_autograd_vgr, 'T': anp_T, 'data': anp_data},
                    {'testName': 'value_and_grad jax', 'fun': dummyLikelihood_jax_vgr, 'T': jnp_T, 'data': jnp_data},

                    {'testName': 'value_and_grad autograd', 'fun': dummyLikelihood_autograd_vgr, 'T': anp_T_re, 'data': anp_data_re},
                    {'testName': 'value_and_grad jax', 'fun': dummyLikelihood_jax_vgr, 'T': jnp_T_re, 'data': jnp_data_re},
                ]


for test in functionList:
    time_function(**test)
```

### What jax/jaxlib version are you using?

jax 0.4.8/jaxlib 0.4.7

### Which accelerator(s) are you using?

CPU

### Additional system info

Python 3.10.10; Linux

**Comment by arabusov** [comment]
My results on  Python 3.10.10, Linux 6.2.2, Intel(R) Core(TM) i5-2520M CPU @ 2.50GHz. JAX is three times slower than the autograd version of value_and_gradient.

```
timing likelihood numpy:
dtype is: complex128
average evaluation time (s):     0.2172

timing likelihood autograd:
dtype is: complex128
average evaluation time (s):     0.2159

timing likelihood jax:
dtype is: complex128
average evaluation time (s):     0.1312

timing likelihood numpy:
dtype is: float64
average evaluation time (s):     0.0930

timing likelihood autograd:
dtype is: float64
average evaluation time (s):     0.0926

timing likelihood jax:
dtype is: float64
average evaluation time (s):     0.0426

timing value_and_grad autograd:
dtype is: complex128
average evaluation time (s):     0.3901

timing value_and_grad jax:
dtype is: complex128
average evaluation time (s):     0.9244

timing value_and_grad autograd:
dtype is: float64
average evaluation time (s):     0.1719

timing value_and_grad jax:
dtype is: float64
average evaluation time (s):     0.0849
```
